### PR TITLE
Add param converters for authentication credentials

### LINKED ISSUE #1373: Invalid credentials should result in 401
## What happened?
When a client makes a request with invalid credentials, they receive a 400 response. This is especially problematic because conjure clients do not propagate 400s, meaning these request ultimately manifest as 500s in the calling service.

## What did you want to happen?
Invalid credentials should result in a 401 response.

This was the outcome of the discussion in https://github.com/palantir/auth-tokens/pull/228. This was implemented in conjure-undertow in https://github.com/palantir/conjure-java/pull/656.

### PR DESCRIPTION
**Pull request by pkoenig10**
## Before this PR
Malformed authentication credentials will result in a 400. This is unfortunate because it will propagate as a 500.

## After this PR
Malformed authentication credentials will result in a 401.

This change is the equivalent to part of https://github.com/palantir/conjure-java/pull/656 for Jersey servers.

### CODE REVIEW

File: conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/AuthHeaderParamConverterProvider.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                ErrorType.Code.UNAUTHORIZED,
+                "Conjure:MissingCredentials");
+        private static final ErrorType MALFORMED_CREDENTIAL_ERROR_TYPE = ErrorType.create(
+                ErrorType.Code.UNAUTHORIZED,
+                "Conjure:MalformedCredentials");
+
+        @Override
+        public AuthHeader fromString(final String value) {
+            if (value == null) {
+                throw new UnauthorizedException(MISSING_CREDENTIAL_ERROR_TYPE);
```
Comment (pkoenig10) [inline conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/AuthHeaderParamConverterProvider.java]: This needs to be a `WebApplicationException`. Otherwise it will get wrapped in a `ExtractorException` which will get mapped to a 400:
https://github.com/jersey/jersey/blob/2.25.1/core-server/src/main/java/org/glassfish/jersey/server/internal/inject/AbstractParamValueExtractor.java#L137

The jaxrs `NotAuthorizedException` doesn't provide constructors that can be used in this scenario so we need to create our own. I chose the name `UnauthorizedException` to disambiguate.

File: conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/GuavaOptionalParamConverterProvider.java [hunk 2]
```diff
@@ -24,10 +24,12 @@
 import javax.ws.rs.ext.ParamConverterProvider;
 import javax.ws.rs.ext.Provider;
 import org.glassfish.hk2.api.ServiceLocator;
+import org.glassfish.jersey.internal.inject.Custom;
 import org.glassfish.jersey.internal.inject.Providers;
 import org.glassfish.jersey.internal.util.ReflectionHelper;
 import org.glassfish.jersey.internal.util.collection.ClassTypePair;
 
+@Custom
```
Comment (pkoenig10) [inline conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/GuavaOptionalParamConverterProvider.java:33]: This annotation ensures that our custom param converters are considered first. 
Otherwise the order in which the param converter providers are considered is arbitrary:
https://github.com/jersey/jersey/blob/2.25.1/core-server/src/main/java/org/glassfish/jersey/server/internal/inject/ParamConverterFactory.java#L58-L101

For example, here is an order that was used when I was debugging tests locally:
param-converter-providers

This is especially important for the `AuthHeaderParamConverterProvider` and `BearerTokenParamConverterProvider` to ensure that we don't fallback to the built-in `AggregatedProvider` which contains the `TypeValueOf` param converter (which is used currently for `AuthHeader` and `BearerToken`).

File: conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import javax.ws.rs.ext.ParamConverter;
+import javax.ws.rs.ext.ParamConverterProvider;
+import javax.ws.rs.ext.Provider;
+import org.glassfish.jersey.internal.inject.Custom;
+
+@Custom
+@Provider
+public final class BearerTokenParamConverterProvider implements ParamConverterProvider {
```
Comment (carterkozak) [inline conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java:34]: I'm not confident this is correct, while cookie auth uses the BearerToken type, we can define additional parameters of type BearerToken that aren't the conjure-defined auth token.

Comment (pkoenig10) [reply conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java:34]: I mention this in the PR description:
> These param converters will apply to all `AuthHeader` and `BearerToken` parameters. However it is very rare for these types to be used outside of the endpoint auth. And most of those cases are to perform some sort of delegate authentication, so a 401 is not inappropriate.

I don't think I've ever seen a `BearerToken` used as an argument outside of cookie auth. And the only times I've seen a `AuthHeader` used as argument outside of header auth is for delegating authorization, in which case you could argue that a 401 is not inappropriate.

I don't think there is another way to achieve the desired behavior. If you have other suggestions, I'm happy to hear them. But I think the benefit of responding with 401 for invalid auth credentials outweighs the minor downside of this applying to other parameters of the same type.

Comment (pkoenig10) [reply conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java:34]: I've updated this PR to only apply to Conjure auth parameters. This means:
- `AuthHeader` parameter must be annotated with `@HeaderParam` and have a value of `Authorization`
- `BearerToken` parameter must be annotated with `@CookieParam` and have any value

Comment (carterkozak) [reply conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java:34]: That sounds reasonable, I don't think we need the additional specificity for AuthHeader because conjure only uses that type for auth components, bearer tokens use the `BearerToken` type.

Comment (pkoenig10) [reply conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java:34]: I would prefer to not respond with 401 for parameters we know aren't the Conjure auth parameter. It's possible to have other `AuthHeader` header params by using external imports. We actually do this for a number of endpoints and use a second `AuthHeader` header param to represent a delegating user. It makes more sense to return 400 for these parameters since they are effectively arguments that just happen to be passed as headers.

Because Conjure doesn't allow arbitrary cookie parameters, this implementation will exactly match all auth parameters and nothing else.

File: conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/AuthHeaderParamConverterProvider.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                ? (ParamConverter<T>) paramConverter
+                : null;
+    }
+
+    public static final class AuthHeaderParamConverter implements ParamConverter<AuthHeader> {
+        @Override
+        public AuthHeader fromString(final String value) {
+            if (value == null) {
+                throw UnauthorizedException.missingCredentials();
+            }
```
Comment (carterkozak) [inline conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/AuthHeaderParamConverterProvider.java:54]: I think this is a break, currently conjure-jersey allows null auth token values. It shouldn't, which is why we fixed it in conjure-undertow, however there are a few projects which rely (or at one point relied) on tokens being nullable.

I am generally supportive of reducing friction migrating between frameworks, but we need to be careful about rollout.

**Comment by pkoenig10** [comment]
@carterkozak are there any other changes we should make in this PR to try to gracefully handle applications that rely on tokens being nullable?

**Comment by carterkozak** [comment]
Let's add test coverage to make sure this change doesn't modify null token behavior, this way we can decouple the risky piece from the type of exception thrown from our infrastructure. We can hold off on null handling until the office opens up and we can discuss more broadly :-)

File: conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        @Override
+        public String toString(final BearerToken value) {
+            Preconditions.checkArgument(value != null);
+            return value.toString();
+        }
+    }
+
+    private static boolean hasAuthAnnotation(Annotation[] annotations) {
+        for (Annotation annotation : annotations) {
+            if (annotation.annotationType() == CookieParam.class) {
```
Comment (carterkozak) [inline conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/BearerTokenParamConverterProvider.java]: nit:
```suggestion
            if (CookieParam.class.equals(annotation.annotationType())) {
```

File: conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/AuthHeaderParamConverterProvider.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        @Override
+        public String toString(final AuthHeader value) {
+            Preconditions.checkArgument(value != null);
+            return value.toString();
+        }
+    }
+
+    private static boolean hasAuthAnnotation(Annotation[] annotations) {
+        for (Annotation annotation : annotations) {
+            if (annotation.annotationType() == HeaderParam.class) {
```
Comment (carterkozak) [inline conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/AuthHeaderParamConverterProvider.java]: nit:
```suggestion
            if (HeaderParam.class.equals(annotation.annotationType())) {
```

File: conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/AuthHeaderParamConverterProvider.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                String value = ((HeaderParam) annotation).value();
+                if (value.equals(HttpHeaders.AUTHORIZATION)) {
```
Comment (carterkozak) [inline conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/AuthHeaderParamConverterProvider.java]: nit, safer around nulls:
```suggestion
                if (HttpHeaders.AUTHORIZATION.equals(value)) {
```

File: conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/GuavaOptionalParamConverterProvider.java (hunk 2 above)
Comment (carterkozak) [reply conjure-java-jersey-server/src/main/java/com/palantir/conjure/java/server/jersey/GuavaOptionalParamConverterProvider.java:33]: Something along these lines would make an excellent comment in the code :-)

**Review by carterkozak — COMMENTED** [review]
Generally looks good to me, thanks! Just a couple small nits

**Comment by carterkozak** [comment]
Thanks! 👍
